### PR TITLE
Zhihan/debugging mode

### PR DESCRIPTION
**Pull request by ScarletGuo**
### Evaluation
* added try exception around precision/recall computing
* returned list of evaluation stats for the purpose of auto recording in batch runs
### Data Loading
* added .lower() to column 'value' when loading clean dataset in order to match the normalization process in loading dirty datasets
* fixed formatting error in table.py's constructor
### Repair
* stored the weights of featurizers after training in repair model
* return the above in the format of string
* optionally print the string above with a newly added flag --printfw

### CODE REVIEW

File: holoclean.py [hunk 1]
```diff
@@ -214,16 +219,22 @@ def repair_errors(self, featurizers):
             print('Time to collect inferred values: %.2f secs' % time)
         status, time = self.ds.get_repaired_dataset()
         print(status)
+        if self.env['verbose']:
```
Comment (thodrek) [inline holoclean.py]: Shouldn't we be checking for the 'printfw' flag (instead of verbose)?

File: repair/featurize/featurize.py [hunk 2]
```diff
@@ -13,6 +13,8 @@ def __init__(self, dataset, env, featurizers):
         for f in featurizers:
             f.setup_featurizer(self.ds, self.total_vars, self.classes, self.processes)
         tensors = [f.create_tensor() for f in featurizers]
+        # save size info of all featurizers
+        self.featurizer_info = [(str(type(featurizers[i])), t.cpu().numpy().shape[2]) for i, t in enumerate(tensors)]
```
Comment (thodrek) [inline repair/featurize/featurize.py]: why isn't t.size() enough? I don't think we need t.cpu().numpy().shape[2] we can just use t.size() (similar to shape for numpy arrays)

File: repair/featurize/occurattrfeat.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            #Get tuple from raw_dataset
+            tid = row['_tid_']
+            tuple = self.raw_data_dict[tid]
+            feat_tensor = self.gen_feat_tensor(row, tuple)
+            tensors.append(feat_tensor)
+        combined = torch.cat(tensors)
+        return combined
+
+    def gen_feat_tensor(self, row, tuple):
+        tensor = torch.zeros(1, self.classes, self.attrs_number*self.attrs_number)
```
Comment (thodrek) [inline repair/featurize/occurattrfeat.py:58]: we should generate a attrs_number*(attrs_number-1) tensor. Otherwise we have |attrs_number| features that will always be zero and hence just occupy unnecessary memory.

Comment (thodrek) [reply repair/featurize/occurattrfeat.py:58]: This will require editing how the tensor is populated to ensure correct alignment.

Comment (thodrek) [reply repair/featurize/occurattrfeat.py:58]: Do we need the extra zero fields for alignment or interpretability or something? Thoughts? @minafarid

**Review by thodrek — CHANGES_REQUESTED** [review]
Please go over the comments. Some need to be addressed.

File: repair/featurize/featurize.py (hunk 2 above)
Comment (ScarletGuo) [reply repair/featurize/featurize.py]: changed

File: holoclean.py (hunk 1 above)
Comment (ScarletGuo) [reply holoclean.py]: changed, but not sure if we need print time info. so i put verbose checking instead printfw checking and print time when verbose is true.

Comment (thodrek) [reply holoclean.py]: In this case, where is the 'printfw' flag used?

**Review by thodrek — COMMENTED** [review]
The changes look good to me. Just waiting for @minafarid to check the featurizer.

File: evaluate/eval.py [hunk 4]
```diff
@@ -70,11 +70,14 @@ def eval_report(self):
             prec, rec, rep_recall, f1, rep_f1 = self.evaluate_repairs()
             report = "Precision = %.2f, Recall = %.2f, Repairing Recall = %.2f, F1 = %.2f, Repairing F1 = %.2f, Detected Errors = %d, Total Errors = %d, Correct Repairs = %d, Total Repairs = %d, Total Repairs (Grdth present) = %d" % (
                       prec, rec, rep_recall, f1, rep_f1, self.detected_errors, self.total_errors, self.correct_repairs, self.total_repairs, self.total_repairs_grdt)
+            report_list = [prec, rec, rep_recall, f1, rep_f1, self.detected_errors, self.total_errors,
+                           self.correct_repairs, self.total_repairs, self.total_repairs_grdt]
         except Exception as e:
             report = "ERROR generating evaluation report: %s"%str(e)
+            report_list = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
Comment (minafarid) [inline evaluate/eval.py]: I think -1 is better NA values for accuracy

File: evaluate/eval.py [hunk 5]
```diff
@@ -148,13 +151,25 @@ def compute_correct_repairs(self):
         self.correct_repairs = correct_repairs
 
     def compute_recall(self):
-        return self.correct_repairs / self.total_errors
+        try:
+            return self.correct_repairs / self.total_errors
+        except Exception as e:
```
Comment (minafarid) [inline evaluate/eval.py]: what could go wrong here? Division by zero?
If so, please use `except ZeroDivisionError as e:`  instead. Or better, check that `self.total_errors` is not zero first and if it is zero, handle it by returning 0 or 1 or whatever you see fit in that case

Comment (minafarid) [reply evaluate/eval.py]: Same for the rest of the compute quality methods

File: repair/featurize/occurattrfeat.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        rv_attr_idx = self.ds.attr_to_idx[rv_attr]
+        for attr in self.all_attrs:
+            if attr != rv_attr and (not pd.isnull(tuple[attr])):
+                attr_idx = self.ds.attr_to_idx[attr]
+                val = tuple[attr]
+                count1 = float(self.single_stats[attr][val])
+                # Get topK values
+                if val not in self.pair_stats[attr][rv_attr]:
+                    if not pd.isnull(tuple[rv_attr]):
+                        raise Exception('Something is wrong with the pairwise statistics. <Val> should be present in dictionary.')
```
Comment (minafarid) [inline repair/featurize/occurattrfeat.py:72]: Can you print which `val` does not appear in the stats, e.g., change it to the following: 
`'Cannot find attribute: %s with value %s in pair-wise statistics' % (attr, val)`

File: repair/featurize/occurattrfeat.py (hunk 3 above)
Comment (minafarid) [reply repair/featurize/occurattrfeat.py:58]: Doesn't the masking process do that already to pad tensors for alignment? I might be missing some details here.

File: holoclean.py [hunk 7]
```diff
@@ -104,7 +104,12 @@
         {'default': False,
          'dest': 'bias',
          'action': 'store_true',
-         'help': 'Use bias term'})
+         'help': 'Use bias term'}),
+    (tuple(['--printfw']),
+        {'default': False,
+         'dest': 'print_feat_weights',
```
Comment (minafarid) [inline holoclean.py]: Could we change `print_feat_weights`  to `printfw` similar to all the other arguments that have the same dest name as the flag name?

**Review by minafarid — CHANGES_REQUESTED** [review]

File: dataset/table.py [hunk 8]
```diff
@@ -31,9 +31,8 @@ def __init__(self, name, src, *args):
                 # Normalize to lower strings and strip whitespaces.
                 # TODO: No support for numerical values. To be added.
                 for attr in self.df.columns.values:
-                    if attr != '_tid_' and self.df[attr].dtype == str:
-                        self.df[attr] = self.df[attr].apply(lambda x: x.lower())
-                        self.df[attr] = self.df[attr].apply(lambda x: x.strip())
+                    if attr != '_tid_':
+                        self.df[attr] = self.df[attr].apply(lambda x: x.lower().strip() if type(x) == str else x)
```
Comment (minafarid) [inline dataset/table.py]: might conflict with https://github.com/HoloClean/holoclean/pull/18/files?utf8=%E2%9C%93&diff=split&w=1#diff-c25c4677d626699f60b8cf3ab8235f4dR51
Use the other PR as a guidance to change this if conflicts

File: holoclean.py (hunk 1 above)
Comment (ScarletGuo) [reply holoclean.py]: it's due to inconsistent naming in flag name and dest name as Mina suggested. I've changed both of them to printfw

File: evaluate/eval.py (hunk 4 above)
Comment (ScarletGuo) [reply evaluate/eval.py]: changed

File: evaluate/eval.py (hunk 5 above)
Comment (ScarletGuo) [reply evaluate/eval.py]: Yes, division by zero. I changed them to the latter.

File: holoclean.py (hunk 7 above)
Comment (ScarletGuo) [reply holoclean.py]: changed both to printfw

File: repair/featurize/occurattrfeat.py (hunk 6 above)
Comment (ScarletGuo) [reply repair/featurize/occurattrfeat.py:72]: i will make changes according to latest occurfeat in dev branch

File: evaluate/eval.py [hunk 9]
```diff
@@ -151,25 +151,22 @@ def compute_correct_repairs(self):
         self.correct_repairs = correct_repairs
 
     def compute_recall(self):
-        try:
+        if self.total_errors == 0:
+            return 0
+        else:
```
Comment (minafarid) [inline evaluate/eval.py]: no need for else. Just return the division result. same for other functions

**Review by minafarid — APPROVED** [review]

File: evaluate/eval.py [hunk 10]
```diff
@@ -70,11 +70,14 @@ def eval_report(self):
             prec, rec, rep_recall, f1, rep_f1 = self.evaluate_repairs()
             report = "Precision = %.2f, Recall = %.2f, Repairing Recall = %.2f, F1 = %.2f, Repairing F1 = %.2f, Detected Errors = %d, Total Errors = %d, Correct Repairs = %d, Total Repairs = %d, Total Repairs (Grdth present) = %d" % (
                       prec, rec, rep_recall, f1, rep_f1, self.detected_errors, self.total_errors, self.correct_repairs, self.total_repairs, self.total_repairs_grdt)
+            report_list = [prec, rec, rep_recall, f1, rep_f1, self.detected_errors, self.total_errors,
+                           self.correct_repairs, self.total_repairs, self.total_repairs_grdt]
         except Exception as e:
```
Comment (richardwu) [inline evaluate/eval.py:86]: Can you modify this exception handling to be similar to the new idiom? 
https://github.com/HoloClean/holoclean/pull/23

i.e. do not silence it but log an error message and re-raise the error.

File: holoclean.py [hunk 11]
```diff
@@ -216,14 +221,22 @@ def repair_errors(self, featurizers):
         print(status)
         if self.env['verbose']:
             print('Time to store repaired dataset: %.2f secs' % time)
+        if self.env['print_fw']:
+            status, time = self.repair_engine.get_featurizer_weights()
+            print(status)
```
Comment (richardwu) [inline holoclean.py]: You'll need to update these `print` statements to using `logging` as in https://github.com/HoloClean/holoclean/pull/25.

**Review by richardwu — CHANGES_REQUESTED** [review]

File: evaluate/eval.py (hunk 10 above)
Comment (ScarletGuo) [reply evaluate/eval.py:86]: raised exception. shall I log info after merging the latest pull request about logging to avoid conflicts

Comment (richardwu) [reply evaluate/eval.py:86]: Yes that should be fine.

Comment (richardwu) [reply evaluate/eval.py:86]: FYI the exception PR has been merged into master https://github.com/HoloClean/holoclean/blob/master/evaluate/eval.py#L72-L76. You'll need to rebase this ontop of master.

File: repair/featurize/occurattrfeat.py (hunk 3 above)
Comment (ScarletGuo) [reply repair/featurize/occurattrfeat.py:58]: is that okay here or changes needed?

File: evaluate/eval.py [hunk 12]
```diff
@@ -77,12 +77,18 @@ def evaluate_repairs(self):
...
-                  prec, rec, rep_recall, f1, rep_f1, self.detected_errors, self.total_errors, self.correct_repairs, self.total_repairs, self.total_repairs_grdt)
+        try:
+            prec, rec, rep_recall, f1, rep_f1 = self.evaluate_repairs()
+            report = "Precision = %.2f, Recall = %.2f, Repairing Recall = %.2f, F1 = %.2f, Repairing F1 = %.2f, Detected Errors = %d, Total Errors = %d, Correct Repairs = %d, Total Repairs = %d, Total Repairs (Grdth present) = %d" % (
+                      prec, rec, rep_recall, f1, rep_f1, self.detected_errors, self.total_errors, self.correct_repairs, self.total_repairs, self.total_repairs_grdt)
+            report_list = [prec, rec, rep_recall, f1, rep_f1, self.detected_errors, self.total_errors,
+                           self.correct_repairs, self.total_repairs, self.total_repairs_grdt]
+        except Exception as e:
+            logging.error("ERROR generating evaluation report %s" % e)
+            raise Exception("ERROR generating evaluation report")
```
Comment (richardwu) [inline evaluate/eval.py]: You should re-raise this instead of raising another exception, otherwise the stack trace will be destroyed, i.e.
```
raise
```
with no arguments (since you've already logged the error message).

**Review by thodrek — APPROVED** [review]
